Configuration of runShadow task appears to be broken by Gradle 6.4.x

### Shadow Version

5.2.0

### Gradle Version

6.4.1

### Expected Behavior

The `runShadow` task requires that `mainClassName` be set (currently after the application of the `application` plug-in and before the application of the `shadow` plug-in).  For example:

```
apply plugin: 'application'
mainClassName = 'io.vantiq.vertx.launcher.VantiqLauncher'
apply plugin: 'com.github.johnrengelman.shadow'
```

Normally when applied in this manner the application runs as expected.  This has been true for all tested versions of Gradle 5.x and is true for Gradle 6.3 (and we presume all prior 6.x releases, though we have not explicitly tested this).

### Actual Behavior

Doing this in Gradle 6.4.1 produces the error:

```
$ ./gradlew servers:standalone:runShadow
> Task :servers:standalone:runShadow FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':servers:standalone:runShadow'.
> The value for task ':servers:standalone:runShadow' property 'mainClass' is final and cannot be changed any further.
```

If I had to guess I'd say this was collateral damage for the changes relating to Java Module support.  In particular you can see that the definition of the JavaApplication type changed from [6.3](https://docs.gradle.org/6.3/dsl/org.gradle.api.plugins.JavaApplication.html) to [6.4.1](https://docs.gradle.org/current/dsl/org.gradle.api.plugins.JavaApplication.html).  Note the removal of `mainClassName` and the introduction of `mainClass`.

This seems like an extremely unfriendly thing to do in a "minor" release and it may well have been unintentional.  But it appears to have completely broken this task.  I can't find a way around this.  Setting `mainClass` in the `application` block as the Gradle documentation suggests leads to the error:

```
* What went wrong:
A problem was found with the configuration of task ':servers:standalone:shadowJar' (type 'ShadowJar').
> No value has been specified for property 'mainClassName'.
```

I have been unable to find something that both sets that value and doesn't violate the fact that it can't be set.

--- a/src/main/groovy/com/github/jengelman/gradle/plugins/shadow/ShadowApplicationPlugin.groovy
+++ b/src/main/groovy/com/github/jengelman/gradle/plugins/shadow/ShadowApplicationPlugin.groovy
@@ -5,7 +5,6 @@ import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
 import org.gradle.api.GradleException
 import org.gradle.api.Plugin
 import org.gradle.api.Project
-import org.gradle.api.Task
 import org.gradle.api.distribution.Distribution
 import org.gradle.api.distribution.DistributionContainer
 import org.gradle.api.file.CopySpec
@@ -71,6 +70,7 @@ class ShadowApplicationPlugin implements Plugin<Project> {
         def run = project.tasks.create(SHADOW_RUN_TASK_NAME, JavaJarExec)
         Sync install = project.tasks.getByName(SHADOW_INSTALL_TASK_NAME)
         run.dependsOn SHADOW_INSTALL_TASK_NAME
+        run.setMain('-jar')
         run.description  = 'Runs this project as a JVM application using the shadow jar'
         run.group = ApplicationPlugin.APPLICATION_GROUP
         run.conventionMapping.jvmArgs = { pluginConvention.applicationDefaultJvmArgs }

--- a/src/main/groovy/com/github/jengelman/gradle/plugins/shadow/internal/JavaJarExec.groovy
+++ b/src/main/groovy/com/github/jengelman/gradle/plugins/shadow/internal/JavaJarExec.groovy
@@ -2,7 +2,6 @@ package com.github.jengelman.gradle.plugins.shadow.internal
 
 import org.gradle.api.tasks.InputFile
 import org.gradle.api.tasks.JavaExec
-import org.gradle.api.tasks.Optional
 import org.gradle.api.tasks.TaskAction
 
 class JavaJarExec extends JavaExec {
@@ -13,7 +12,6 @@ class JavaJarExec extends JavaExec {
     @Override
     @TaskAction
     public void exec() {
-        setMain('-jar')
         List<String> allArgs = [getJarFile().path] + getArgs()
         setArgs(allArgs)
         super.exec()

